### PR TITLE
Fix early constant folding for isNull/isNotNul and analyzer.

### LINKED ISSUE #64575: [Analyzer] minmax count projection is not used with analyzer
Part of playbench.
```
clickhouse-cloud :) EXPLAIN SELECT count(1) FROM checks WHERE test_name is not null SETTINGS allow_experimental_analyzer=0, allow_experimental_parallel_reading_from_replicas=0

EXPLAIN
SELECT count(1)
FROM checks
WHERE test_name IS NOT NULL
SETTINGS allow_experimental_analyzer = 0, allow_experimental_parallel_reading_from_replicas = 0

Query id: 499bc5b7-98a1-40fa-b640-e2baa3ca7f26

   ┌─explain─────────────────────────────────────────────────┐
1. │ Expression ((Projection + Before ORDER BY))             │
2. │   Aggregating                                           │
3. │     Expression                                          │
4. │       ReadFromPreparedSource (_minmax_count_projection) │
   └─────────────────────────────────────────────────────────┘

4 rows in set. Elapsed: 0.006 sec.

clickhouse-cloud :) EXPLAIN SELECT count(1) FROM checks WHERE test_name is not null SETTINGS allow_experimental_analyzer=1, allow_experimental_parallel_reading_from_replicas=0

EXPLAIN
SELECT count(1)
FROM checks
WHERE test_name IS NOT NULL
SETTINGS allow_experimental_analyzer = 1, allow_experimental_parallel_reading_from_replicas = 0

Query id: 62bf5a49-b694-419e-a590-9e8e67748f59

   ┌─explain────────────────────────────────────────────────────────────┐
1. │ Expression ((Project names + Projection))                          │
2. │   Aggregating                                                      │
3. │     Expression (Before GROUP BY)                                   │
4. │       Filter ((WHERE + Change column names to column identifiers)) │
5. │         ReadFromMergeTree (checks.checks)                          │
   └────────────────────────────────────────────────────────────────────┘

5 rows in set. Elapsed: 0.002 sec.

clickhouse-cloud :) SELECT count(1) FROM checks WHERE test_name is not null SETTINGS allow_experimental_analyzer=1, allow_experimental_parallel_reading_from_replicas=0

SELECT count(1)
FROM checks
WHERE test_name IS NOT NULL
SETTINGS allow_experimental_analyzer = 1, allow_experimental_parallel_reading_from_replicas = 0

Query id: 0fc52f3d-75a3-49a6-96c7-1cfe8c5744a7

   ┌───count(1)─┐
1. │ 3121567112 │ -- 3.12 billion
   └────────────┘

1 row in set. Elapsed: 26.163 sec. Processed 3.12 billion rows, 6.24 GB (119.31 million rows/s., 238.63 MB/s.)
Peak memory usage: 806.02 MiB.

clickhouse-cloud :) SELECT count(1) FROM checks WHERE test_name is not null SETTINGS allow_experimental_analyzer=0, allow_experimental_parallel_reading_from_replicas=0

SELECT count(1)
FROM checks
WHERE test_name IS NOT NULL
SETTINGS allow_experimental_analyzer = 0, allow_experimental_parallel_reading_from_replicas = 0

Query id: 696750a6-f29e-461d-b6f2-846a08533dd5

   ┌────count()─┐
1. │ 3121567112 │ -- 3.12 billion
   └────────────┘

1 row in set. Elapsed: 0.004 sec.

SHOW CREATE TABLE checks

CREATE TABLE checks.checks
(
    `pull_request_number` UInt32,
    `commit_sha` LowCardinality(String),
    `check_name` LowCardinality(String),
    `check_status` LowCardinality(String),
    `check_duration_ms` UInt64,
    `check_start_time` DateTime,
    `test_name` LowCardinality(String),
    `test_status` LowCardinality(String),
    `test_duration_ms` UInt64,
    `report_url` String,
    `pull_request_url` String,
    `commit_url` String,
    `task_url` String,
    `base_ref` String,
    `base_repo` String,
    `head_ref` String,
    `head_repo` String,
    `test_context_raw` String,
    `instance_type` LowCardinality(String),
    `instance_id` String,
    `date` Date MATERIALIZED toDate(check_start_time)
)
ENGINE = SharedMergeTree('/clickhouse/tables/{uuid}/{shard}', '{replica}')
PRIMARY KEY (date, pull_request_number, commit_sha, check_name, test_name, check_start_time)
ORDER BY (date, pull_request_number, commit_sha, check_name, test_name, check_start_time)
SETTINGS index_granularity = 8192
```

### PR DESCRIPTION
**Pull request by KochetovNicolai**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix removing the `WHERE` and `PREWHERE` expressions, which are always true (for the new analyzer)

Fixes  #64575

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit c9ce1362a19fd344403100d1fca2351e9ec700b4 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64695/c9ce1362a19fd344403100d1fca2351e9ec700b4/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by novikd — APPROVED** [review]

File: src/DataTypes/DataTypeNullable.cpp [hunk 1]
```diff
@@ -176,16 +176,8 @@ DataTypePtr removeNullableOrLowCardinalityNullable(const DataTypePtr & type)
...
 
 bool canContainNull(const IDataType & type)
-    {
-        if (type.isNullable() || type.isLowCardinalityNullable() || isDynamic(type))
-            return true;
-
-        if (const auto * variant = typeid_cast<const DataTypeVariant *>(&type))
-            for (const auto & elem : variant->getVariants())
-                if (canContainNull(*elem))
-                    return true;
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/DataTypes/DataTypeNullable.cpp]: No. Looks like variant just always can be nullable, this is implemented in selector.

File: src/Planner/PlannerExpressionAnalysis.cpp [hunk 2]
```diff
@@ -539,8 +539,11 @@ PlannerExpressionsAnalysisResult buildExpressionAnalysisResult(const QueryTreeNo
     if (query_node.hasWhere())
     {
         where_analysis_result_optional = analyzeFilter(query_node.getWhere(), current_output_columns, planner_context, actions_chain);
-        where_action_step_index_optional = actions_chain.getLastStepIndex();
-        current_output_columns = actions_chain.getLastStepAvailableOutputColumns();
+        if (where_analysis_result_optional)
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Planner/PlannerExpressionAnalysis.cpp:542]: We are returning std::nullopt if where condition is constant 1.

**Comment by clickgapai** [comment]
Hi @KochetovNicolai @novikd — while reviewing this PR I found the following:

- **Bug (P2):** `isNullable` returns 0 for constant Dynamic/Variant but 1 for non-constant - inconsistent constant folding — https://github.com/ClickHouse/ClickHouse/issues/104506

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**